contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,476 @@
+a number of objects of different kinds) should be `similar` to lists in core
+API (by lists I mean e.g. `PodList` object)
+- the API has to be implemented also in aggregator so that bulk operations
+are supported also if different resource types are served by different
+apiservers
+- clients has to be able to alter their watch subscribtions incrementally (it
+may not be implemented in the initial version though, but has to be designed)
+
+
+# Detailed design
+
+ As stated in above requirements, we need to make bulk operations work across
+different resource types (e.g. watch pod P and secret S within a single watch
+call). Spanning multiple resources, resource types or conditions will be more
+and more important for large number of watches. As an example, deferation will
+be adding watches for every type it federates. With that in mind, bypassing
+aggregation at the resource type level and going to aggregation over objects
+with different resource types will allow us to more aggresively optimize in the
+future (it doesn't mean you have to watch resources of different types in a
+single watch, but we would like to make it possible).
+
+ That means, that we need to implement the bulk operation at the aggregation
+level. The implications of it are discussed below.
+
+ Moreover, our current REST API doesn't even offer an easy way to handle
+"multiple watches of a given type" within a single request. As a result, instead
+of inventing new top level pattern per type, we can introduce a new resource
+type that follows normal RESTful rules and solves even more generic problem
+of spanning multiple different resource types.
+
+ We will start with introducing a new dedicated API group:
+ ```
+ /apis/bulk.k8s.io/
+ ```
+ that underneath will have a completely separate implementation.
+
+ In all text below, we are assuming v1 version of the API, but it will obviously
+go through alpha and beta stages before (it will start as v1alpha1).
+
+ In this design, we will focus only on bulk get (list) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `getoperation` resource and supporting the
+following operation:
+```
+POST /apis/bulk.k8s.io/v1/getoperation <body defines filtering>
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+
+ We could consider adding `watch` operation using the same pattern with just
+`?watch=1` parameter. However, the main drawback of this approach is that it
+won't allow for dynamic altering of watch subscriptions (which we definitely
+also need to support).
+As a result, we need another API for watch that will also support incremental
+subscriptions - it will look as following:
+```
+websocket /apis/bulk.k8s.io/v1/getoperation?watch=1
```

Comment:
In line 98 above I have:
"In all text below, we are assuming v1 version of the API, but it will obviously go through alpha and beta stages before (it will start as v1alpha1)."